kvpress/presses/kvzip_press.py

```diff
@@ -0,0 +1,358 @@
+            chunk_num = (ctx_len - 1) // chunk_size + 1
+
+            input_ids = []
+            for i in range(chunk_num):
+                start = i * chunk_size
+                end = (i + 1) * chunk_size
+                a_ids = ctx_ids[:, start:end]
+                if a_ids.shape[1] == 0:
+                    continue
+                input_ids.append(a_ids)
+        else:
+            input_ids = [ctx_ids]
+
+        return input_ids
+
+    def prepare(
+        self,
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizer,
+        chunk_size: int = 2048,
+        prev_postfix_size=8,
+    ):
+        """Prepare chunked inputs for KV importance scoring with context reconstruction
+        return: List[torch.Tensor]
+        """
+        ctx_ids = self._context_ids[:, self._prefix_len :].to("cpu")
+
+        # initialize score values
+        n_layers, n_kv = model.config.num_hidden_layers, model.config.num_key_value_heads
+        self.score_val = torch.zeros(
+            (n_layers, 1, n_kv, self.context_length),  # only support batch size of 1
+            dtype=model.dtype,
+            device=model.device,
+        )
+        self.score_val[..., : self.n_sink] = 1.0
+
+        input_ids = []
+        chunked_inputs = self._chunk_fn(ctx_ids, chunk_size)
+        for i, a_ids in enumerate(chunked_inputs):
+            if i == 0:
+                prompt = "\n\nRepeat the previous context exactly."
+                q_ids = tokenizer.encode(prompt, return_tensors="pt", add_special_tokens=False)
+            else:
+                prompt = "\n\nRepeat the part of the previous context exactly, starting with"
+                q_ids = tokenizer.encode(prompt, return_tensors="pt", add_special_tokens=False)
+                postfix_prev = chunked_inputs[i - 1][:, -prev_postfix_size:]
+                q_ids = torch.cat([q_ids, postfix_prev], dim=1)
+
+            input_ids.append((a_ids, torch.cat([q_ids, self._suffix_ids, a_ids], dim=1)))
+
+        return input_ids
+
+    def _make_mask(self, attn_weights: torch.Tensor, window_size: int):
+        """
+        Define causal mask shared across layers
+        """
+        mask = torch.full((window_size, window_size), torch.finfo(attn_weights.dtype).min, device=attn_weights.device)
+        mask_cond = torch.arange(mask.size(-1), device=attn_weights.device)
+        mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+        self.causal_mask_score = mask[None, None, None, :, :]
+
+    def _mask_causal(self, attn_weights: torch.Tensor, window_size: int):
+        """
+        Apply causal maksing
+        """
+        if self.causal_mask_score is None:
+            self._make_mask(attn_weights, window_size)
+        elif self.causal_mask_score.size(-1) != window_size:
+            self._make_mask(attn_weights, window_size)
+
+        attn_weights[..., -window_size:, -window_size:] += self.causal_mask_score
+
+    def score(
```

Comment:
This method name is a bit confusing, as `.score` is already used in ScorerPresses, but with a different signature.